src/tactic/itauto.lean

```diff
@@ -0,0 +1,486 @@
+| (prop.imp (prop.and ak A B) C) p Γ :=
+  let (A, B) := ak.sides A B in
+  Γ.add (prop.imp A (B.imp C)) (p.curry ak A B)
+| (prop.imp (prop.or A B) C) p Γ := do
+  Γ ← Γ.add (A.imp C) (p.or_imp_left A B),
+  Γ.add (B.imp C) (p.or_imp_right A B)
+| (prop.imp A prop.true) p Γ := pure Γ
+| A p Γ := pure (Γ.insert A p)
+
+/-- Add `A` to the context `Γ` with proof `p`. This version of `context.add` takes a continuation
+and a target proposition `B`, so that in the case that `⊥` is found we can skip the continuation
+and just prove `B` outright. -/
+@[inline] meta def context.with_add {m} [monad m] (Γ : context) (A : prop) (p : proof)
+  (B : prop) (f : context → prop → m proof) : m proof :=
+match Γ.add A p with
+| except.ok Γ_A := f Γ_A B
+| except.error p := pure (p B)
+end
+
+/-- The search phase, which deals with the level 3 rules, which are rules that are not validity
+preserving and so require proof search. One obvious one is the or-introduction rule: we prove
+`A ∨ B` by proving `A` or `B`, and we might have to try one and backtrack.
+
+There are two rules dealing with implication in this category: `p, p -> C ⊢ B` where `p` is an
+atom (which is safe if we can find it but often requires the right search to expose the `p`
+assumption), and `(A₁ → A₂) → C ⊢ B`. We decompose the double implication into two subgoals: one to
+prove `A₁ → A₂`, which can be written `A₂ → C, A₁ ⊢ A₂` (where we used `A₁` to simplify
+`(A₁ → A₂) → C`), and one to use the consequent, `C ⊢ B`. The search here is that there are
+potentially many implications to split like this, and we have to try all of them if we want to be
+complete. -/
+meta def search (prove : context → prop → state_t ℕ option proof) :
+  prop → context → state_t ℕ option proof
+| B Γ := match Γ.find B with
+  | some p := pure p
+  | none := match B with
+    | prop.or B₁ B₂ := proof.or_inl B₂ <$> prove Γ B₁ <|> proof.or_inr B₁ <$> prove Γ B₂
+    | _ := ⟨λ n, Γ.fold none $ λ A p r, r <|> match A with
+      | prop.imp A' C := match Γ.find A' with
+        | some q := (context.with_add (Γ.erase A) C (p.app q) B prove).1 n
+        | none := match A' with
+          | prop.imp A₁ A₂ :=
+            (do {
+              let Γ : context := Γ.erase A,
+              a ← fresh_name,
+              p₁ ← Γ.with_add A₁ (proof.hyp a) A₂ $ λ Γ_A₁ A₂,
+                Γ_A₁.with_add (prop.imp A₂ C) (proof.imp_imp_simp a A₁ A₂ p) A₂ prove,
+              Γ.with_add C (p.app (proof.intro a A₁ p₁)) B prove
+            } : state_t ℕ option proof).1 n
+          | _ := none
+          end
+        end
+      | _ := none
+      end⟩
+    end
+  end
+
+/-- The main prover. This receives a context of proven or assumed lemmas and a target proposition,
+and returns a proof or `none` (with state for the fresh variable generator).
+The intuitionistic logic rules are separated into three groups:
+
+* level 1: No splitting, validity preserving: apply whenever you can.
```

Comment:
Can you put this description in the module doc, i.e., at the beginning?  The other docstrings refer to this classification as well.